src/lib/y2firewall/widgets/pages/zone.rb

```diff
@@ -72,60 +74,92 @@ def label
+            sctp: N_("SCTP Ports"),
+            # TRANSLATORS: DCCP is the Datagram Congestion Control Protocol
+            dccp: N_("DCCP Ports")
+          }.freeze
 
-        # FIXME: separate objects like this do not work well with the
-        # single zone.ports attribute mixing all protos. Rather make a
-        # CustomWidget that handles it all
-        class PortsForProtocol < CWM::InputField
-          # @param proto [:tcp,:udp,:sctp,:dccp]
-          def initialize(zone, label, proto)
+          def initialize(zone)
+            textdomain "firewall"
             @zone = zone
-            @proto = proto
-            @label = label
-            self.widget_id = "#{proto}_ports"
           end
 
-          attr_reader :label
+          def contents
+            fields = PROTOCOLS.map do |sym, label|
+              InputField(Id(sym), Opt(:hstretch), _(label))
+            end
+            VBox(* fields)
+          end
+
+          def help
+            "FIXME: ports or port ranges, separated by spaces and/or commas <br>" \
+            "a port is an integer <br>" \
+            "a port range is port-dash-port (with no spaces)"
+          end
 
           def init
-            log.info "INIT #{widget_id}"
-            # FIXME: factor out and test
-            self.value = @zone.ports
-                              .map { |p| p.split("/") }
-                              .find_all { |_port, proto| proto == @proto.to_s }
-                              .map { |port, _proto| port }
-                              .join(", ")
+            by_proto = ports_from_array(@zone.ports)
+            PROTOCOLS.each do |sym, _label|
+              Yast::UI.ChangeWidget(Id(sym), :Value, by_proto.fetch(sym, []).join(", "))
+            end
           end
 
+          # FIXME: validation, cleanup, error reporting
           def store
-            # FIXME: do modify immediately?
-            # DUH, clumsy
-            log.info "STORE #{widget_id}"
+            by_proto = PROTOCOLS.map do |sym, _label|
+              line = Yast::UI.QueryWidget(Id(sym), :Value)
+              [sym, items_from_ui(line)]
+            end
+            @zone.ports = ports_to_array(by_proto.to_h)
+          end
+
+        private
+
+          def items_from_ui(s)
+            # the separator is at least one comma or space, surrounded by optional spaces
+            s.split(/ *[, ] */)
+          end
+
+          # @return [String] ports specification as array
+          # @param hash [Hash{Symbol => Array<String>}] ports specification broken by protocol
+          # @example
+          #   h = { tcp: ["55555-55666", "44444"], udp: ["33333"] }
+          #   a = ["55555-55666/tcp", "44444/tcp", "33333/udp"]
+          #   ports_to_array(h) # => a
+          def ports_to_array(hash)
+            hash.map { |sym, ports| ports.map { |p| "#{p}/#{sym}" } }.flatten
+          end
+
+          # @param s [String] ports specification as one string
```

Comment:
param is called `a` instead of `s`. Moreover, I would prefer a more descriptive name :)